### ISSUE TITLE
feature: hook up into and modify query builder before compilation

### ISSUE THREAD

**Issue by fogine** [knex/knex#3060]
# Environment

Knex version: 0.16.3

# Feature discussion / request
Knex currently emits the `start` event once on a top-level (parent) query builder which 
allows to modify the query before its executed, for example appending timestamps etc..
However the event is not run for sub-query builder objects:  

```javascript
return knex('reviews').whereIn('reviews.user_id', function() {
    const q = this.select('id').from('users').where('username', 'test');
    q.on('start', function() {
        //LISTENERS ARE NOT RUN FOR SUBQUERIES
    });

    return q;
}).del();
```
meaning that its not currently possible to modify subqueries before they are executed.  
Its not possible eb.: to append `where deleted_at is null` clause to the above subquery...

There is a drawback in modifying  sub-query builders from `start` event listeners  though being that on certain cases  the `start` event code might be run asynchronously in parallel with the query compilation code which would break things as `EventEmiiter` api is strictly synchronous.

3. Give some API proposal, how the feature should work

To solve the discussed issue, better would be to implement hooks interface as discussed with @elhigu  and allow to asynchronously modify query builders by registering eg. "beforeCompile" hook:  

```javascript
const queryBuilder = knex('reviews');

queryBuilder.whereIn('reviews.user_id', function() {
    const q = this.select('id').from('users').where('username', 'test');

    q.beforeCompile(function(builder) {
        return builder.where('users.deleted_at', null);
    });

    return q;
}).del();

queryBuilder.beforeCompile(function(builder, optionalArg1, optionalArg2/*, ....*/) {
    return builder.where(optionalArg1 + '.deleted_at', null);
}, 'reviews', 'optionalArg2Value');

return queryBuilder;
```

**Comment by elhigu** [comment]
This is more like a feature request... IIRC `start` event is currently emitted once for query... But since it is done before compilation phase I don't see why it could not be emitted for each builder... 

However actually that being an event is kind of bad thing, since it will mean that we might be running an event code and query compilation code parallel and it would  mean that by the time when event code is finished query might have been already compiled and sent to db driver.

So better feature would be to provide async hooks beforeCompile,  which would be executed during compiling phase.

Could you rephrase this issue to be a feature request, with API suggestion etc. ?

**Comment by mercmobily** [comment]
Thank you for raising this about hooks to modify query builders before compilation. As elhigu noted, this is indeed a feature request rather than a bug report.

The current start event fires once per query before compilation. Even though it technically allows builder modification, elhigu correctly identified the critical flaw: events create a race condition risk. If event handlers were async, they could run parallel to compilation, potentially causing the query to be compiled and sent to the database before the event handler completes.

As elhigu explained, using events for this purpose is fundamentally problematic. Events don't guarantee execution order with async operations, and by the time event code finishes, the query might already be compiled and sent to the driver. This makes the current start event unsuitable for reliable query modification.

This issue overlaps significantly with #3632, which requests hooks for observability and debugging. Even though that issue focuses on monitoring and metrics rather than query modification, both needs would be served by the same underlying hook system. As elhigu noted in #3632, the current events are "quite a mess" and new well-designed hooks could deprecate older patterns.

For query modification specifically, elhigu proposed async hooks that execute during the compilation phase (not parallel to it), suggesting beforeCompile hooks. This would ensure hooks complete before compilation proceeds, eliminating race conditions and providing proper async support with predictable execution order.

Looking at the broader picture, there are multiple related issues requesting similar functionality:

* Issue #3632 requests observability hooks with Raynos specifically needing query-pre-acquire and query-post-acquire events to measure true client-side latency including connection pool wait times. The current events only fire after connection acquisition, missing critical timing information.

* Issue #6173 discusses query augmentation for multi-tenant applications, needing automatic WHERE clause injection for tenant isolation.

* Issue #2789 requested callbacks before query execution for validation, with elhigu suggesting the start event as a workaround even though it has limitations.

* Issue #4823 and #6221 both deal with request-scoped query filters and Row Level Security patterns that would benefit from proper hooks.

As elhigu requested, this needs to be rephrased as a proper feature request with clear API proposals, specification of hook execution order and timing, examples of how hooks would integrate with the compilation phase, and error handling behaviour.

Since both this issue and #3632 essentially need the same underlying hook system, and given that the current issues lack proper structure and API proposals, I recommend closing both this issue and #3632 in favour of a new, unified feature request (#6270)  that addresses all hook needs comprehensively.

The new issue should present a complete hook system design that covers both query modification use cases (multi-tenant filtering, soft deletes, query validation) and observability use cases (timing, metrics, tracing, debugging). It should include concrete API proposals showing how hooks integrate into the query lifecycle, from before connection acquisition through to after result processing. The design should address elhigu's concerns about race conditions by ensuring hooks execute sequentially during their respective phases rather than parallel to them. It should also consider Raynos's need for earlier hook points to capture connection pool wait times and tmc's suggestion about async context propagation.

This unified approach would avoid fragmenting the discussion across multiple issues and provide a single, well-designed hook system that serves all use cases, allowing deprecation of the current "messy" event system as elhigu suggested.

Related issues that should be referenced in the new unified feature request:
- #3632 - Hooks for observability and debugging
- #6173 - Query augmentation patterns
- #2789 - Callback before query execution
- #4823 - Request scoped query filters  
- #6221 - Centralized Query-Level RLS
- #5987 - Query middleware/interceptors
